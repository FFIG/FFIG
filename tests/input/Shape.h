#include <cmath>

// Any class annotated with this macro will be exposed and have all functions
// exposed.
// Any class deriving from an exposed class will have its constructors exposed.
#ifdef __clang__
#define C_API __attribute__((annotate("GENERATE_C_API")))
#else
#define C_API
#endif

struct AbstractShape
{
  virtual ~AbstractShape()
  {
  }
  virtual double area() const = 0;
  virtual double perimeter() const = 0;
  virtual const char* name() const = 0;
  virtual int is_equal(const AbstractShape* s) const = 0;
} C_API;

static const double pi = 3.14159265359;

class Circle : public AbstractShape
{
  const double radius_;

public:
  double area() const override
  {
    return pi * radius_ * radius_;
  }

  double perimeter() const override
  {
    return 2 * pi * radius_;
  }

  const char* name() const override
  {
    return "Circle";
  }

  int is_equal(const AbstractShape* s) const override
  {
    if ( auto c = dynamic_cast<const Circle*>(s) )
      return c->radius_ == radius_;
    return false;
  }

  Circle(double radius) : radius_(radius)
  {
    if ( radius < 0 ) 
    { 
      std::string s = "Circle radius \"" + std::to_string(radius_) + "\" must be non-negative.";
      throw std::runtime_error(s);
    }
  }
};

class Square : public AbstractShape
{
  const double side_;

public:
  double area() const override
  {
    return side_ * side_;
  }

  double perimeter() const override
  {
    return 4.0 * side_;
  }

  const char* name() const override
  {
    return "Square";
  }

  int is_equal(const AbstractShape* s) const override
  {
    if ( auto sq = dynamic_cast<const Square*>(s) )
      return sq->side_ == side_;
    return false;
  }

  Square(double side) : side_(side)
  {
    if ( side_ <= 0.0 )
    {
      std::string s = "Square side \"" + std::to_string(side_) + "\" must be non-negative.";
      throw std::runtime_error(s);
    }
  }
};

class Pentagon : public AbstractShape
{
  const double side_;

public:
  double area() const override
  {
    return 0.25 * sqrt(5. * (5. + 2. * sqrt(5.))) * side_ * side_;
  }

  double perimeter() const override
  {
    return 5.0 * side_;
  }

  const char* name() const override
  {
    return "Pentagon";
  }

  int is_equal(const AbstractShape* s) const override
  {
    if ( auto p = dynamic_cast<const Pentagon*>(s) )
      return p->side_ == side_;
    return false;
  }

  Pentagon(double side) : side_(side)
  {
    if ( side_ <= 0.0 )
    {
      std::string s = "Pentagon side \"" + std::to_string(side_) + "\" must be non-negative.";
      throw std::runtime_error(s);
    }
  }
};
