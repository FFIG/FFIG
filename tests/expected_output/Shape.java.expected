// This code was generated by FFIG <http://ffig.org>.
// Manual edits will be lost.
package org.ffig.Shape

import com.sun.jna.Library;
import com.sun.jna.Native;
import com.sun.jna.Pointer;
import com.sun.jna.ptr.PointerByReference;
import com.sun.jna.ptr.DoubleByReference;
import com.sun.jna.ptr.IntByReference;

interface ShapeCLibrary extends Library
{
    ShapeCLibrary INSTANCE = (ShapeCLibrary)Native.loadLibrary(
        "libShape_c.dylib", ShapeCLibrary.class);

    void Shape_clear_error();

    String Shape_error();

    int Shape_AbstractShape_dispose(Pointer o);

    int Shape_AbstractShape_area(Pointer o, DoubleByReference rv);

    int Shape_AbstractShape_perimeter(Pointer o, DoubleByReference rv);

    int Shape_AbstractShape_name(Pointer o, PointerByReference rv);

    int Shape_AbstractShape_is_equal(Pointer o, Pointer s, IntByReference rv);

    int Shape_Circle_create(double radius, PointerByReference o);

    int Shape_Square_create(double side, PointerByReference o);

    int Shape_Pentagon_create(double side, PointerByReference o);
}

class ShapeException extends RuntimeException {
    public ShapeException()
    {
        super(ShapeCLibrary.INSTANCE.Shape_error());
        ShapeCLibrary.INSTANCE.Shape_clear_error();
    }
}

public class AbstractShape {
    private Pointer ptr = Pointer.NULL;

    // Pointer constructor needs to be package-access for factory methods.
    AbstractShape(Pointer ptr)
    {
        this.ptr = ptr;
    }

    protected void finalize() {
        ShapeCLibrary.INSTANCE.Shape_AbstractShape_dispose(ptr);
    }

    public double area() {
      DoubleByReference rv = new DoubleByReference();;
      int rc = ShapeCLibrary.INSTANCE.Shape_AbstractShape_area(ptr, rv);
      if(rc != 0) {
        throw new ShapeException();
      }
      return rv.getValue();
    }

    public double perimeter() {
      DoubleByReference rv = new DoubleByReference();;
      int rc = ShapeCLibrary.INSTANCE.Shape_AbstractShape_perimeter(ptr, rv);
      if(rc != 0) {
        throw new ShapeException();
      }
      return rv.getValue();
    }

    public String name() {
      PointerByReference rv = new PointerByReference();;
      int rc = ShapeCLibrary.INSTANCE.Shape_AbstractShape_name(ptr, rv);
      if(rc != 0) {
        throw new ShapeException();
      }
      return rv.getValue().getString(0);
    }

    public int is_equal(AbstractShape s) {
      IntByReference rv = new IntByReference();;
      int rc = ShapeCLibrary.INSTANCE.Shape_AbstractShape_is_equal(ptr, s.ptr, rv);
      if(rc != 0) {
        throw new ShapeException();
      }
      return rv.getValue();
    }

// Use static factory methods for sub-classes as Java does not allow multiple 
// public classes in a file.

    public static AbstractShape AbstractShape_Circle_create(double radius) {
        PointerByReference rv = new PointerByReference();
        int rc = ShapeCLibrary.INSTANCE.Shape_Circle_create(radius, rv);
        if(rc != 0) {
            throw new ShapeException();
        }
        return new AbstractShape(rv.getValue());
    }

    public static AbstractShape AbstractShape_Square_create(double side) {
        PointerByReference rv = new PointerByReference();
        int rc = ShapeCLibrary.INSTANCE.Shape_Square_create(side, rv);
        if(rc != 0) {
            throw new ShapeException();
        }
        return new AbstractShape(rv.getValue());
    }

    public static AbstractShape AbstractShape_Pentagon_create(double side) {
        PointerByReference rv = new PointerByReference();
        int rc = ShapeCLibrary.INSTANCE.Shape_Pentagon_create(side, rv);
        if(rc != 0) {
            throw new ShapeException();
        }
        return new AbstractShape(rv.getValue());
    }
}
